azurelinuxagent/ga/cgroupapi.py

```diff
@@ -103,28 +109,84 @@ def _foreach_legacy_cgroup(operation):
                     procs_file_contents = fileutil.read_file(procs_file).strip()
-                    daemon_pid = CGroupsApi.get_daemon_pid()
+                    daemon_pid = CGroupUtil.get_daemon_pid()
 
                     if ustr(daemon_pid) in procs_file_contents:
                         operation(controller, daemon_pid)
         finally:
             for _, cgroup in legacy_cgroups:
-                logger.info('Removing {0}', cgroup)
+                log_cgroup_info('Removing {0}'.format(cgroup), send_event=False)
                 shutil.rmtree(cgroup, ignore_errors=True)
         return len(legacy_cgroups)
 
     @staticmethod
-    def get_daemon_pid():
-        return int(fileutil.read_file(get_agent_pid_file_path()).strip())
+    def cleanup_legacy_cgroups():
+        """
+        Previous versions of the daemon (2.2.31-2.2.40) wrote their PID to /sys/fs/cgroup/{cpu,memory}/WALinuxAgent/WALinuxAgent;
+        starting from version 2.2.41 we track the agent service in walinuxagent.service instead of WALinuxAgent/WALinuxAgent. If
+        we find that any of the legacy groups include the PID of the daemon then we need to disable data collection for this
+        instance (under systemd, moving PIDs across the cgroup file system can produce unpredictable results)
+        """
+        return CGroupUtil._foreach_legacy_cgroup(lambda *_: None)
 
 
-class SystemdCgroupsApi(CGroupsApi):
+class SystemdRunError(CGroupsException):
+    """
+    Raised when systemd-run fails
+    """
+
+    def __init__(self, msg=None):
+        super(SystemdRunError, self).__init__(msg)
+
+
+def get_cgroup_api():
     """
-    Cgroups interface via systemd
+    Determines which version of Cgroup should be used for resource enforcement and monitoring by the Agent and returns
+    the corresponding Api.
+
+    Uses 'stat -f --format=%T /sys/fs/cgroup' to get the cgroup hierarchy in use.
+        If the result is 'cgroup2fs', cgroup v2 is being used.
+        If the result is 'tmpfs', cgroup v1 or a hybrid mode is being used.
+            If the result of 'stat -f --format=%T /sys/fs/cgroup/unified' is 'cgroup2fs', then hybrid mode is being used.
+
+    Raises exception if cgroup filesystem mountpoint is not '/sys/fs/cgroup', or an unknown mode is detected. Also
+    raises exception if hybrid mode is detected and there are controllers available to be enabled in the unified
+    hierarchy (the agent does not support cgroups if there are controllers simultaneously attached to v1 and v2
+    hierarchies).
     """
+    if not os.path.exists(CGROUP_FILE_SYSTEM_ROOT):
+        v1_mount_point = shellutil.run_command(['findmnt', '-t', 'cgroup', '--noheadings'])
+        v2_mount_point = shellutil.run_command(['findmnt', '-t', 'cgroup2', '--noheadings'])
+        raise CGroupsException("Expected cgroup filesystem to be mounted at '/sys/fs/cgroup', but it is not.\n v1 mount point: {0}\n v2 mount point: {1}".format(v1_mount_point, v2_mount_point))
```

Comment:
Minor: for v1 it will be a list of mountpoints, let's add a newline v1 mount point: **\n**{0}\n. Maybe the same for v2 for consistency.